# Comparison of vanilla Python vs `numpy`

Below, a series of experiments are run to demonstrate the difference in speed between Python and numpy. The comparisons will be on the following categories:

- Basic Tests
  - generating an array
  - summing up the elements of an array
  - taking the average of the elements of an array
- Computing Non-trivial Functions
  - sorting the elements of an array
  - computing the standard deviation of the elements of an array
  - computing the sine of every element in the array
  - computing the exponential of every element in the array
- Manipulating Multiple Arrays
  - computing element-wise sum over two arrays
  - computing element-wise multiplication over two arrays
  - taking the dot product of two arrays

To conduct these tests, three sets of functions will be used:

- `naive`: a naïve approach, using vanilla Python lists and `for` loops
- `compr`: using Python __list comprehension__, a very useful (and optimized) version of Python's basic `for` loops, and other built-in functions (that are also optimized)
- `numpy`: using the `numpy` library and its versatile set of functions

In [1]:
import numpy as np
import math
import random
import operator

In [2]:
ARRAY_SIZE = 10000

## Basic Tests

In [3]:
def generate_naive(fill=1, size=ARRAY_SIZE):
    naive_list = []
    for i in range(size):
        naive_list.append(fill)
    return naive_list

def generate_compr(fill=1, size=ARRAY_SIZE):
    compr_list = [fill] * size
    return compr_list

def generate_numpy(fill=1, size=ARRAY_SIZE):
    numpy_arr = np.full((size,), fill)
    return numpy_arr

In [4]:
def generate_rand_naive(size=ARRAY_SIZE):
    naive_list = []
    for i in range(size):
        naive_list.append(random.random())
    return naive_list

def generate_rand_compr(size=ARRAY_SIZE):
    compr_list = [random.random() for i in range(size)]
    return compr_list

def generate_rand_numpy(size=ARRAY_SIZE):
    numpy_arr = np.random.rand(size)
    return numpy_arr

In [5]:
py_rand = generate_rand_compr()
np_rand = np.array(py_rand)

In [6]:
print('Generating via naive for loop, filled with 1, size', ARRAY_SIZE)
%timeit generate_naive()
print('Generating via list comprehension, filled with 1, size', ARRAY_SIZE)
%timeit generate_compr()
print('Generating via numpy.array(), filled with 1, size', ARRAY_SIZE)
%timeit generate_numpy()

Generating via naive for loop, filled with 1, size 10000
797 µs ± 8.54 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Generating via list comprehension, filled with 1, size 10000
24.4 µs ± 493 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Generating via numpy.array(), filled with 1, size 10000
7.99 µs ± 158 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [7]:
print('Generating via naive for loop, filled with random values, size', ARRAY_SIZE)
%timeit generate_rand_naive()
print('Generating via list comprehension, filled with random values, size', ARRAY_SIZE)
%timeit generate_rand_compr()
print('Generating via numpy.array(), filled with random values, size', ARRAY_SIZE)
%timeit generate_rand_numpy()

Generating via naive for loop, filled with random values, size 10000
1.96 ms ± 80.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Generating via list comprehension, filled with random values, size 10000
1.29 ms ± 16 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Generating via numpy.array(), filled with random values, size 10000
68 µs ± 664 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [8]:
def sum_naive(arr):
    tot = 0
    for el in arr:
        tot += el
    return tot

def sum_compr(arr):
    return sum(arr)

def sum_numpy(arr):
    return np.sum(arr)

In [9]:
print('Summing via naive for loop, size', ARRAY_SIZE)
%timeit sum_naive(py_rand)
print('Summing via built-in sum(), size', ARRAY_SIZE)
%timeit sum_compr(py_rand)
print('Summing via numpy.sum(), size', ARRAY_SIZE)
%timeit sum_numpy(np_rand)

Summing via naive for loop, size 10000
445 µs ± 8.54 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Summing via built-in sum(), size 10000
55.1 µs ± 583 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Summing via numpy.sum(), size 10000
10.4 µs ± 1.29 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [10]:
def avg_naive(arr):
    tot = 0
    cnt = 0
    for el in arr:
        tot += el
        cnt += 1
    return tot / cnt

def avg_compr(arr):
    tot = sum(arr)
    cnt = len(arr)
    return tot / cnt

def avg_numpy(arr):
    avg = np.mean(arr)
    return avg

In [11]:
print('Averaging via naive for loop, size', ARRAY_SIZE)
%timeit avg_naive(py_rand)
print('Averaging via built-in sum() and len(), size', ARRAY_SIZE)
%timeit avg_compr(py_rand)
print('Averaging via numpy.mean(), size', ARRAY_SIZE)
%timeit avg_numpy(np_rand)

Averaging via naive for loop, size 10000
1.42 ms ± 23.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Averaging via built-in sum() and len(), size 10000
54.5 µs ± 958 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
Averaging via numpy.mean(), size 10000
13.5 µs ± 115 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


### Analysis

Below is a table summarizing the results (all arrays of size $10000$, times all in $\mu s$):

| task | naive | compr | `numpy` |
| :- | -: | -: | -: |
| generate array of 1s | $797 \pm 8.54$ | $24.4 \pm 0.493$ | $7.99 \pm 0.158$ |
| generate random array | $1960 \pm 80.9$ | $1290 \pm 16$ | $68 \pm 0.664$ |
| compute sum | $445 \pm 8.54$ | $55.1 \pm 0.583$ | $10.4 \pm 1.29$ |
| compute average | $1420 \pm 23.9$ | $54.5 \pm 0.958$ | $13.5 \pm 0.115$ |

Here, we can see that for almost all of the tasks, list comprehension is one order of magnitude faster than using naive `for` loops, and using `numpy` is one more order of magnitude faster than that. The only exception is the case of generating arrays filled with random values. Here, we see that list comprehension still offers a decent speed boost, well beyond what is statistically significant. The reason why list comprehension is slower here is that unlike the other three tasks, we cannot "hide" the for loop. If we were to try a similar method as generating an array of $1$s (e.g. `[random.random()] * size`), the resultant array will have the correct size, and the values will be random, but they will all be the same!

Almost without exception, *implicit* for loops are faster than *explicit* ones (although in some cases, code readability may be more important). This is because Python's built-in functions are written using the C programming language, which is much more verbose (technical term here is "strictly typed"). The verbosity makes it __harder to play around with the data__, but makes the code __run much faster__, because the computer can apply a bunch of tricks specifically designed for the explicit type of data being used. Python is designed to be easy to write and experiment with, and for most tasks, it is fast __enough__ since the tasks don't get repeated thousands or millions of times. However, when the task does need to be repeated, Python offers alternative solutions that are optimized behind the scenes which makes the code go *zoom zoom*.

## Computing Non-trivial Functions

In [12]:
def sort_naive(arr, lo, hi):
    # simple implementation of the well-known quicksort
    def swap(arr, i, j):
        temp = arr[i]
        arr[i] = arr[j]
        arr[j] = temp
    
    def partition(arr, lo, hi):
        pivot = arr[hi]
        i = lo
        for j in range(lo, hi):
            if arr[j] < pivot:
                swap(arr, i, j)
                i += 1
        swap(arr, i, hi)
        return i
    
    if lo < hi:
        p = partition(arr, lo, hi)
        sort_naive(arr, lo, p-1)
        sort_naive(arr, p+1, hi)

def sort_compr(arr):
    arr.sort() # note that this calls Python's built-in sort() function

def sort_numpy(arr):
    arr.sort() # note that this calls numpy's built-in sort() function

In [13]:
print('Sorting via naive quicksort, size', ARRAY_SIZE)
%timeit sort_naive(generate_rand_naive(), 0, ARRAY_SIZE-1)
print('Sorting via built-in sort(), size', ARRAY_SIZE)
%timeit sort_compr(generate_rand_compr())
print('Sorting via numpy.sort(), size', ARRAY_SIZE)
%timeit sort_numpy(generate_rand_numpy())

Sorting via naive quicksort, size 10000
58.5 ms ± 1.07 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
Sorting via built-in sort(), size 10000
2.33 ms ± 100 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Sorting via numpy.sort(), size 10000
541 µs ± 6.53 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [14]:
def std_naive(arr):
    avg = avg_naive(arr)
    square_diff = 0
    cnt = 0
    for el in arr:
        square_diff += (el - avg) ** 2
        cnt += 1
    return math.sqrt(square_diff / cnt)

def std_compr(arr):
    avg = avg_compr(arr)
    square_diff = sum([(val - avg)**2 for val in arr])
    return math.sqrt(square_diff / len(arr))

def std_numpy(np_arr):
    std = np.std(np_arr)
    return std

In [15]:
print('Standard deviation via naive loop, size', ARRAY_SIZE)
%timeit std_naive(py_rand)
print('Standard deviation via list comprehension, size', ARRAY_SIZE)
%timeit std_compr(py_rand)
print('Standard deviation via np.std(), size', ARRAY_SIZE)
%timeit std_numpy(np_rand)

Standard deviation via naive loop, size 10000
4.03 ms ± 39.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Standard deviation via list comprehension, size 10000
1.43 ms ± 17.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Standard deviation via np.std(), size 10000
40.7 µs ± 544 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [16]:
def sin_naive(arr):
    sin_arr = []
    for el in arr:
        sin_arr.append(math.sin(el))
    return sin_arr

def sin_compr(arr):
    sin_arr = list(map(math.sin, py_rand))
    return sin_arr

def sin_numpy(arr):
    sin_arr = np.sin(arr)
    return sin_arr

In [17]:
print('Trigonmetric sine via naive loop, size', ARRAY_SIZE)
%timeit sin_naive(py_rand)
print('Trigonmetric sine via built-in map(), size', ARRAY_SIZE)
%timeit sin_compr(py_rand)
print('Trigonmetric sine via np.sin(), size', ARRAY_SIZE)
%timeit sin_numpy(np_rand)

Trigonmetric sine via naive loop, size 10000
2.14 ms ± 21 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Trigonmetric sine via built-in map(), size 10000
770 µs ± 6.68 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Trigonmetric sine via np.sin(), size 10000
129 µs ± 731 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [18]:
def exp_naive(arr):
    exp_arr = []
    for el in arr:
        exp_arr.append(math.e ** el)
    return exp_arr

def exp_compr(arr):
    # this is also more accurate!
    exp_arr = list(map(math.exp, py_rand))
    return exp_arr

def exp_numpy(arr):
    exp_arr = np.exp(arr)
    return exp_arr

In [19]:
print('Exponential via naive loop, size', ARRAY_SIZE)
%timeit exp_naive(py_rand)
print('Exponential via built-in map(), size', ARRAY_SIZE)
%timeit exp_compr(py_rand)
print('Exponential via np.exp(), size', ARRAY_SIZE)
%timeit exp_numpy(np_rand)

Exponential via naive loop, size 10000
2.43 ms ± 13.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
Exponential via built-in map(), size 10000
726 µs ± 3.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Exponential via np.exp(), size 10000
124 µs ± 2.27 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


### Analysis

Below is the table summary for this section (array size is 10000, time units in $\mu s$):

| task | naive | compr | `numpy` |
| :- | -: | -: | -: |
| sort (raw) | $5850 \pm 1070$ | $2330 \pm 100$ | $541 \pm 6.53$ |
| sort (clean) | $3890 \pm 1150$ | $1040 \pm 120$ | $473 \pm 7.18$ |
| std. dev. | $4030 \pm 39.8$ | $1430 \pm 17.9$ | $40.7 \pm 0.544$ |
| sin(x) | $2140 \pm 21.0$ | $770 \pm 6.68$ | $129 \pm 0.731$ |
| exp(x) | $2430 \pm 13.1$ | $726 \pm 3.50$ | $124 \pm 2.27$ |

Here, we see again that `numpy` is 100 times faster than the naive approach. However, we see that list comprehension is no longer a 10x improvement over the naive approach, and is instead close to only 3 times faster. This is primarily because these complicated functions require more individual computations (e.g. Taylor expansions), and the difference between `numpy` and the built-in functions adds up.

For the sorting task, I ran the sorting algorithm on a new, randomly generated array every single time to average out the strengths and weaknesses of various sorting algorithms. The "raw" row shows the total time measured, while the "clean" row excludes the time needed to run the random generation, as measured in the previous section. Note that for this task, the built-in function performs very well when compared to `numpy`; this is because the same sorting algorithm is used, and the only hidden function being run is the comparison function (i.e. is element A smaller than element B?), which gives us the opposite result compared to the other three functions.

## Manipulating Multiple Arrays

In [20]:
py_secn = generate_rand_compr()
np_secn = np.array(py_secn)

In [21]:
def sum_mult_naive(arr1, arr2):
    sum_arr = []
    for i in range(len(arr1)):
        sum_arr.append(arr1[i] + arr2[i])
    return sum_arr

def sum_mult_compr(arr1, arr2):
    sum_arr = list(map(operator.add, arr1, arr2))
    return sum_arr
    # sum_arr = [a1 + a2 for a1, a2 in zip(arr1, arr2)]
    # while the above line is okay, it's actually about twice as slow as the other one

def sum_mult_numpy(arr1, arr2):
    sum_arr = np.add(arr1, arr2) # can also use arr1 + arr2
    return sum_arr

In [22]:
print('Sum of two python lists via naive for loop, size', ARRAY_SIZE)
%timeit sum_mult_naive(py_rand, py_secn)
print('Sum of two python lists via built-in map(), size', ARRAY_SIZE)
%timeit sum_mult_compr(py_rand, py_secn)
print('Sum of two numpy arrays via np.add(), size', ARRAY_SIZE)
%timeit sum_mult_numpy(np_rand, np_secn)

Sum of two python lists via naive for loop, size 10000
1.59 ms ± 52.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Sum of two python lists via built-in map(), size 10000
414 µs ± 5.82 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Sum of two numpy arrays via np.add(), size 10000
5.05 µs ± 27.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [23]:
def prod_mult_naive(arr1, arr2):
    prod_arr = []
    for i in range(len(arr1)):
        prod_arr.append(arr1[i] * arr2[i])
    return prod_arr

def prod_mult_compr(arr1, arr2):
    prod_arr = list(map(operator.mul, arr1, arr2))
    return prod_arr
    # prod_arr = [a1 * a2 for a1, a2 in zip(arr1, arr2)]

def prod_mult_numpy(arr1, arr2):
    prod_arr = np.multiply(arr1, arr2) # can also use arr1 * arr2
    return prod_arr

In [24]:
print('Product of two python lists via naive for loop, size', ARRAY_SIZE)
%timeit prod_mult_naive(py_rand, py_secn)
print('Product of two python lists via built-in map(), size', ARRAY_SIZE)
%timeit prod_mult_compr(py_rand, py_secn)
print('Product of two numpy arrays via np.multiply(), size', ARRAY_SIZE)
%timeit prod_mult_numpy(np_rand, np_secn)

Product of two python lists via naive for loop, size 10000
1.56 ms ± 17.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Product of two python lists via built-in map(), size 10000
433 µs ± 2.57 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Product of two numpy arrays via np.multiply(), size 10000
5.12 µs ± 111 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [25]:
def dot_prod_naive(arr1, arr2):
    dp = 0
    for i in range(len(arr1)):
        dp += arr1[i] * arr2[i]
    return dp

def dot_prod_compr(arr1, arr2):
    dp = sum(map(operator.mul, arr1, arr2))
    return dp

def dot_prod_numpy(arr1, arr2):
    dp = np.dot(arr1, arr2)
    return dp

In [26]:
print('Dot product of two python lists via naive for loop, size', ARRAY_SIZE)
%timeit prod_mult_naive(py_rand, py_secn)
print('Dot product of two python lists via built-in map(), size', ARRAY_SIZE)
%timeit prod_mult_compr(py_rand, py_secn)
print('Dot product of two numpy arrays via numpy.dot(), size', ARRAY_SIZE)
%timeit prod_mult_numpy(np_rand, np_secn)

Dot product of two python lists via naive for loop, size 10000
1.74 ms ± 62.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Dot product of two python lists via built-in map(), size 10000
442 µs ± 15.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
Dot product of two numpy arrays via numpy.dot(), size 10000
5.46 µs ± 78 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


### Analysis
One more table, array size is 10000, times in $\mu s$:

| task | naive | compr | `numpy` |
| :- | -: | -: | -: |
| element-wise sum | $1590 \pm 52.7$ | $414 \pm 5.82$ | $5.05 \pm 0.027$ |
| element-wise prod | $1560 \pm 17.3$ | $433 \pm 2.57$ | $5.12 \pm 0.111$ |
| dot product | $1740 \pm 62.7$ | $442 \pm 15.8$ | $5.46 \pm 0.078$ |

Here, we see that `numpy` performs a whopping *300 times* faster than the basic for loop, and almost 100 times better than list comprehension. This shows where `numpy` really starts to shine: with multiple arrays and multi-dimensional arrays. The reason for this is discussed in the appendix, but the general gist is that `numpy` stores its arrays inside the computer memory differently from how Python handles lists. This reflects the differences in design philosophies (as discussed earlier), but the best part is that using Python and its plethora of libraries, you are able to get the best of both worlds.


# Appendix
Many useful libraries such as `pandas`, `scipy`, and `scikit-learn` use `numpy` arrays behind the scenes, because it performs so well. One way to check if a particular method uses `numpy` is by checking the type of the result, likeso:

```
result = useful_module.cool_function()
print(type(result))
>>> numpy.ndarray
```

If you see `numpy.ndarray` or something similar, this function probably uses `numpy` in the background.

Note that `numpy` is very finicky about its data types, and more often than not, putting things that aren't numbers into the arrays will cause `numpy` to be unhappy.

** \*\* warning for technical details \*\* **

This is particularly true of Python __objects__, which actually remove one of the core advantages of `numpy`, namely *contiguous memory*. One of the reasons why `numpy` is written in C is the power of working with computer memory directly, and *pointer arithmetic* (which allows code to grab data from arbitrary locations in memory). `numpy` takes advantage of this by storing each array's data in continuous blocks of memory, which reduces both computation time and the memory footprint of the code. However, when Python objects are stored in a `numpy` array, the values inside the array are themselves pointers to the actual memory location of the object. This means that `numpy` is no longer able to use 90% of its fancy tricks, which in turn means that performance degrades *very quickly*.

** \*\* end of technical details \*\* **

Common examples of Python objects that get shoved into `numpy` arrays are `string`s, `list`s, and `tuple`s. If you check the array's data type (use `arr.dtype`) and find that the array is treating your data as objects, try changing the `dtype` property when you initialize the array, or check the documentation for the correct method to store the data format.

However, in the **worst case scenario**, `numpy`'s performance is no worse than using Python lists, which means that you should always use `numpy` arrays over Python lists, whenever possible. Just check to make sure that you're getting the most out of `numpy`'s functionality.